### PR TITLE
Strict resource cache mode

### PR DESCRIPTION
**Pull request by psav**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,16 +3,227 @@
...
 
 ## Resource Cache
-TODO
+### Background Problem Statement
+The Resource Cache is a handy utility for working with k8s resources in situations where you need
+multiple blocks of code to have access to resources before they are written to k8s, without having
+to explicitly pass the objects around. Consider the following, a `Deployment` resource needs to be
+updated by three blocks of code. Traditionally either the resources would need to be passed
+successively to the three blocks of code, or the resource would need to be written to k8s and then
+*got*, and *updated* by each subsequent block.
```
Comment (adamrdrew) [inline README.md:12]: <3 <3 <3 Love the added documentation!

File: README.md [hunk 2]
```diff
@@ -3,16 +3,227 @@
...
+var (
+	scheme   = runtime.NewScheme()
+)
+
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(apps.AddToScheme(scheme))
+}
+
+## Creationg of logger is omitted
```
Comment (adamrdrew) [inline README.md:42]: Spelling on Creation

File: README.md [hunk 3]
```diff
@@ -3,16 +3,227 @@
...
+`ApplyAll()`. In this case, it will be skipped in the `ApplyAll()` step as an `Update()` call will
+immediately update it in k8s. This should be used with care. The object should not be updated after
+the initial update, though there is nothing to prevent this from happening, it ill simply result in
+a second API write.
+
+```go
+NewSingleResourceIdent("prov", "purpose", &core.ConfigMap{}, rc.ResourceOptions{WriteNow: true})
+```
+
+### possibleGVKs and strict mode
```
Comment (adamrdrew) [inline README.md:198]: Should this be moved up to the Create part because it is germane to creation?

File: resource_cache/resource_cache.go [hunk 4]
```diff
@@ -138,13 +138,22 @@ type ObjectCache struct {
 	config          *CacheConfig
 }
 
-func NewCacheConfig(scheme *runtime.Scheme, logKey interface{}, protectedGVKs map[schema.GroupVersionKind]bool, debugOptions DebugOptions) *CacheConfig {
+func NewCacheConfig(scheme *runtime.Scheme, possibleGVKs, protectedGVKs GVKMap, options ...Options) *CacheConfig {
```
Comment (adamrdrew) [inline resource_cache/resource_cache.go:141]: Why do we do this options array and then just pop the first entry off?

File: resource_cache/resource_cache.go [hunk 5]
```diff
@@ -155,12 +164,16 @@ type DebugOptions struct {
...
 
+type Options struct {
+	StrictGVK    bool
+	DebugOptions DebugOptions
+}
+
 type CacheConfig struct {
-	possibleGVKs  map[schema.GroupVersionKind]bool
-	protectedGVKs map[schema.GroupVersionKind]bool
+	possibleGVKs  GVKMap
```
Comment (adamrdrew) [inline resource_cache/resource_cache.go:173]: :+1:

File: resource_cache/resource_cache.go [hunk 6]
```diff
@@ -221,7 +235,14 @@ func (o *ObjectCache) registerGVK(obj client.Object) {
 // blank object is stored in the cache it is imperative that the user of this function call Create
 // before modifying the obejct they wish to be placed in the cache.
 func (o *ObjectCache) Create(resourceIdent ResourceIdent, nn types.NamespacedName, object client.Object) error {
-	o.registerGVK(object)
+	if o.config.options.StrictGVK {
+		gvk, _ := utils.GetKindFromObj(o.scheme, object)
```
Comment (adamrdrew) [inline resource_cache/resource_cache.go:239]: Do we need to handle this error we're underscoring out or is it definitely never gonna pop?

File: resource_cache/resource_cache_test.go [hunk 7]
```diff
@@ -106,22 +110,14 @@ type Key string
...
 	ctx := context.Background()
 	zapLog, _ := zap.NewDevelopment()
 
-	log = zapr.NewLogger(zapLog)
-
-	ctx = context.WithValue(ctx, Key("bunk"), &log)
+	log := zapr.NewLogger(zapLog)
 
-	oCache := NewObjectCache(ctx, k8sClient, &config)
+	oCache := NewObjectCache(ctx, k8sClient, log, config)
```
Comment (adamrdrew) [inline resource_cache/resource_cache_test.go:120]: Weird that we used to pass in `&config` even though `NewObjectCache` expects `*CacheConfig` and `NewCacheConfig` returns `*CacheConfig` - kinda confused why the previous version ever worked.

**Review by adamrdrew — COMMENTED** [review]
Great patch. I noticed no problems, just a couple of questions.

**Review by adamrdrew — APPROVED** [review]
Yeah nothing to say here, this is all good.

File: README.md (hunk 2 above)
Comment (psav) [reply README.md]: damn it aspell

File: README.md (hunk 3 above)
Comment (psav) [reply README.md]: yeh I think so

File: resource_cache/resource_cache.go (hunk 4 above)
Comment (psav) [reply resource_cache/resource_cache.go:141]: because I wanted it to be optional

File: resource_cache/resource_cache.go (hunk 6 above)
Comment (psav) [reply resource_cache/resource_cache.go]: good point

File: resource_cache/resource_cache_test.go (hunk 7 above)
Comment (psav) [reply resource_cache/resource_cache_test.go]: Yeh - just rejigged a little

**Pull request merged**